File: leetcode/go/1371.find-the-longest-substring-containing-vowels-in-even-counts.go
func findTheLongestSubstring(s string) int {
    prefixIdx := make(map[int]int)
    prefixIdx[0] = -1

    mask := 0
    ans := 0
    for i, r := range s {
        if strings.ContainsRune("aeiou", r) {
            mask ^= (1 << (r - 'a'))
        }
        if idx, ok := prefixIdx[mask]; ok {
            ans = max(ans, i - idx)
        } else {
            prefixIdx[mask] = i
        }
    }
    return ans
}

File: leetcode/rust-leetcode/src/1371.find-the-longest-substring-containing-vowels-in-even-counts.rs
use std::collections::HashMap;
impl Solution {
    pub fn find_the_longest_substring(s: String) -> i32 {
        let (mut ans, mut mask) = (0, 0);
        let mut prefixMap = HashMap::from([(0, -1)]);
        for (i, c) in s.chars().enumerate() {
            if "aeiou".contains(c) {
                mask ^= (1 << (c as u8 - b'a'));
            }
            match prefixMap.get(&mask) {
                None => {
                    prefixMap.insert(mask, i as i32);
                },
                Some(idx) => {
                    ans = ans.max(i as i32 - idx);
                }
            }
        }
        ans
    }
}

